File: tests/c/test_vfs_writer_spill.c
/*
 * test_vfs_writer_spill.c — the writer's SPILL mode (phase-107 C1) reorders
 * honestly: absorb any offset into a local scratch, publish only a complete
 * object, and answer capacity with ENOSPC — never a truncated object.
 *
 * WHAT: drives brix_vfs_writer_spill_{enter,put,drain,discard} — the real
 *       vfs_writer_spill.o — through the T1 promotion, out-of-order absorption
 *       with extent coalescing, the sequential T2 drain into a spy staged
 *       session, the T3 discard, and every refusal class (no scratch root,
 *       rewind, overlap, spill_max/T4, coverage hole).
 * WHY:  The reorder buffer is what turns "every reordered upload on http/remote
 *       fails" into a supported configuration; its refusals are load-bearing
 *       security/correctness edges (a hole must never be zero-filled into a
 *       published object, a capacity failure must abort the staged session so
 *       nothing partial publishes, and the scratch must live under the SERVICE
 *       spill root, never the export). A wire test sees only the response;
 *       this unit pins each transition hermetically.
 * HOW:  links the real vfs_writer_spill.o with its whole cross-TU closure as
 *       spies: the backend-registry spill lookup answers a test-controlled
 *       root/cap, the staged session records exactly what the drain streams
 *       (offset-checked, byte-compared), the metric recorders count, and the
 *       temp namer mirrors the owned-temp shape. Real ngx pool objects supply
 *       ngx_palloc/ngx_pnalloc.
 *
 * Cases (success + error + security-negative):
 *   success:      T1 enter flips SEQUENTIAL→SPILL, creates the scratch under
 *                 the spill root, bumps the active gauge; reverse-order puts
 *                 coalesce to one extent run and the drain hands the staged spy
 *                 a byte-exact sequential stream; discard unlinks the scratch,
 *                 drops the gauge, and is idempotent.
 *   error:        no spill root → ENOSPC and the session STAYS SEQUENTIAL; a
 *                 rewind below the staged prefix → EINVAL (stays SEQUENTIAL);
 *                 an overlapping extent → EINVAL (stays SPILL — client error,
 *                 not capacity); a drain over a coverage hole → EINVAL with
 *                 the staged session never touched.
 *   security-neg: spill_max exceeded (at entry and mid-put, T4) aborts the
 *                 staged session WITH temp removal and unlinks the scratch —
 *                 no partial object anywhere — and parks the session FAILED;
 *                 the scratch is created 0600, with the owned-temp reaper name,
 *                 strictly under the spill root and never under the export.
 *
 * Build+run: see tests/cmdscripts/c_object_units.py ("vfs_writer_spill").
 */
#include "fs/vfs/vfs_writer_internal.h"

#include <assert.h>
#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/stat.h>
#include <unistd.h>

/* ---- spy state ---------------------------------------------------------- */
static const char *g_spill_root;     /* what the backend registry answers      */
static off_t       g_spill_max;
static int         g_tmp_counter;

#define SPY_OBJ_MAX  (4u << 20)
static u_char      g_staged_buf[SPY_OBJ_MAX];  /* what the drain streamed      */
static off_t       g_staged_cursor;  /* spy enforces the sequential contract   */
static int         g_staged_writes;
static int         g_staged_aborts;
static unsigned    g_abort_remove_tmp;

static int         g_metric_bytes_calls;
static size_t      g_metric_bytes_sum;
static int         g_metric_refused;
static int         g_metric_active;  /* gauge: sum of the +/-1 deltas          */

static void
reset_spies(void)
{
    g_spill_root         = NULL;
    g_spill_max          = 0;
    memset(g_staged_buf, 0, sizeof(g_staged_buf));
    g_staged_cursor      = 0;
    g_staged_writes      = 0;
    g_staged_aborts      = 0;
    g_abort_remove_tmp   = 0;
    g_metric_bytes_calls = 0;
    g_metric_bytes_sum   = 0;
    g_metric_refused     = 0;
    g_metric_active      = 0;
}

/* ---- cross-TU closure of vfs_writer_spill.o ------------------------------ */

ngx_int_t
brix_vfs_backend_spill(const char *root_canon, const char **root_out,
    off_t *max_out)
{
    (void) root_canon;
    *root_out = g_spill_root;
    *max_out  = g_spill_max;
    return g_spill_root != NULL ? NGX_OK : NGX_ERROR;
}

/* Owned-temp shape (".xrd-tmp.<pid>.") so the reaper-name assertion is real. */
ngx_int_t
brix_make_tmp_path(const char *base_path, char *out, size_t out_sz)
{
    int n = snprintf(out, out_sz, "%s.xrd-tmp.%d.%08x", base_path,
                     (int) getpid(), (unsigned) ++g_tmp_counter);
    return (n > 0 && (size_t) n < out_sz) ? NGX_OK : NGX_ERROR;
}

ngx_int_t
brix_vfs_pwrite_full(ngx_fd_t fd, const u_char *buf, size_t len, off_t off)
{
    while (len > 0) {
        ssize_t n = pwrite(fd, buf, len, off);

        if (n <= 0) {
            return NGX_ERROR;
        }
        buf += n; len -= (size_t) n; off += n;
    }
    return NGX_OK;
}

ngx_int_t
brix_vfs_pread_full(ngx_fd_t fd, u_char *buf, size_t len, off_t off,
    size_t *got)
{
    *got = 0;
    while (*got < len) {
        ssize_t n = pread(fd, buf + *got, len - *got, off + (off_t) *got);

        if (n < 0) {
            return NGX_ERROR;
        }
        if (n == 0) {
            break;                       /* short read: scratch too small */
        }
        *got += (size_t) n;
    }
    return NGX_OK;
}

/* The staged spy IS the sequential contract: an out-of-order drain write here
 * would be exactly the driver-corruption the spill exists to prevent. */
ngx_int_t
brix_vfs_staged_write(brix_vfs_staged_t *st, const void *buf, size_t len,
    off_t off)
{
    (void) st;
    assert(off == g_staged_cursor && "drain must stream sequentially");
    assert((size_t) off + len <= sizeof(g_staged_buf));
    memcpy(g_staged_buf + off, buf, len);
    g_staged_cursor = off + (off_t) len;
    g_staged_writes++;
    return NGX_OK;
}

void
brix_vfs_staged_abort(brix_vfs_staged_t *st, unsigned remove_tmp)
{
    (void) st;
    g_staged_aborts++;
    g_abort_remove_tmp = remove_tmp;
}

void
brix_metric_vfs_spill_bytes(brix_proto_t proto, size_t bytes)
{
    (void) proto;
    g_metric_bytes_calls++;
    g_metric_bytes_sum += bytes;
}

void
brix_metric_vfs_spill_refused(brix_proto_t proto)
{
    (void) proto;
    g_metric_refused++;
}

void
brix_metric_vfs_spill_active(int delta)
{
    g_metric_active += delta;
}

/* ---- harness ------------------------------------------------------------- */

static ngx_log_t       g_log;
static ngx_pool_t     *g_pool;
static brix_vfs_ctx_t  g_ctx;
static char            g_root_dir[128];   /* mkdtemp'd service spill root      */
static char            g_export_dir[160]; /* a sibling "export" for the        */
                                          /* never-under-the-export assertion  */
static int             g_staged_dummy;

static void
writer_init(brix_vfs_writer_t *w, off_t staged_cursor)
{
    memset(w, 0, sizeof(*w));
    w->ctx           = &g_ctx;
    w->pool          = g_pool;
    w->log           = &g_log;
    w->st            = (brix_vfs_staged_t *) &g_staged_dummy;
    w->staged_cursor = staged_cursor;
    w->mode          = BRIX_VFS_WRITER_SEQUENTIAL;
    w->spill.fd      = NGX_INVALID_FILE;
}

static int
scratch_on_disk(const brix_vfs_writer_t *w, struct stat *st)
{
    return w->spill.path != NULL
        && stat((const char *) w->spill.path, st) == 0;
}

/* ---- success ------------------------------------------------------------- */

static void
test_success_enter_creates_scratch_and_promotes(void)
{
    brix_vfs_writer_t w;
    struct stat       st;

    reset_spies();
    g_spill_root = g_root_dir;
    writer_init(&w, 0);
    assert(brix_vfs_writer_spill_enter(&w, 2 << 20, 4096) == NGX_OK);
    assert(w.mode == BRIX_VFS_WRITER_SPILL);
    assert(g_metric_active == 1);
    assert(scratch_on_disk(&w, &st));
    brix_vfs_writer_spill_discard(&w);
}

static void
test_success_reverse_order_drain_is_byte_exact(void)
{
    enum { CHUNK = 64 * 1024, N = 8 };
    static u_char     ref[CHUNK * N];
    brix_vfs_writer_t w;
    int               i;

    reset_spies();
    g_spill_root = g_root_dir;
    for (i = 0; i < (int) sizeof(ref); i++) {
        ref[i] = (u_char) (i * 31 + 7);
    }
    writer_init(&w, 0);
    assert(brix_vfs_writer_spill_enter(&w, (N - 1) * CHUNK, CHUNK) == NGX_OK);
    for (i = N - 1; i >= 0; i--) {          /* strictly reverse order */
        assert(brix_vfs_writer_spill_put(&w, ref + i * CHUNK, CHUNK,
                                         (off_t) i * CHUNK) == NGX_OK);
    }
    assert(w.spill.n_ext == 1 && "exact-touch neighbours must coalesce");
    assert(w.spill.ext[0].start == 0);
    assert(w.spill.ext[0].end == (off_t) sizeof(ref));
    assert(g_metric_bytes_sum == sizeof(ref));

    assert(brix_vfs_writer_spill_drain(&w) == NGX_OK);
    assert(g_staged_cursor == (off_t) sizeof(ref));
    assert(memcmp(g_staged_buf, ref, sizeof(ref)) == 0);
    brix_vfs_writer_spill_discard(&w);
}

static void
test_success_discard_unlinks_and_is_idempotent(void)
{
    brix_vfs_writer_t w;
    struct stat       st;
    char              kept[PATH_MAX];

    reset_spies();
    g_spill_root = g_root_dir;
    writer_init(&w, 0);
    assert(brix_vfs_writer_spill_enter(&w, 4096, 1) == NGX_OK);
    snprintf(kept, sizeof(kept), "%s", (const char *) w.spill.path);
    brix_vfs_writer_spill_discard(&w);
    assert(stat(kept, &st) != 0 && "discard must unlink the scratch");
    assert(g_metric_active == 0);
    brix_vfs_writer_spill_discard(&w);      /* idempotent */
    assert(g_metric_active == 0);
}

/* ---- error ---------------------------------------------------------------- */

static void
test_error_no_spill_root_is_enospc_stays_sequential(void)
{
    brix_vfs_writer_t w;

    reset_spies();
    g_spill_root = NULL;                     /* nothing configured */
    writer_init(&w, 0);
    errno = 0;
    assert(brix_vfs_writer_spill_enter(&w, 4096, 4096) == NGX_ERROR);
    assert(errno == ENOSPC);
    assert(w.mode == BRIX_VFS_WRITER_SEQUENTIAL);
    assert(g_metric_refused == 1);
    assert(g_staged_aborts == 0 && "an in-order continuation must still work");
}

static void
test_error_rewind_below_staged_prefix_is_einval(void)
{
    brix_vfs_writer_t w;

    reset_spies();
    g_spill_root = g_root_dir;
    writer_init(&w, 1 << 20);                /* 1 MiB already staged */
    errno = 0;
    assert(brix_vfs_writer_spill_enter(&w, 4096, 4096) == NGX_ERROR);
    assert(errno == EINVAL);
    assert(w.mode == BRIX_VFS_WRITER_SEQUENTIAL);
    assert(g_metric_refused == 1);
}

static void
test_error_overlap_is_einval_stays_spill(void)
{
    static const u_char blk[4096];
    brix_vfs_writer_t   w;

    reset_spies();
    g_spill_root = g_root_dir;
    writer_init(&w, 0);
    assert(brix_vfs_writer_spill_enter(&w, 8192, sizeof(blk)) == NGX_OK);
    assert(brix_vfs_writer_spill_put(&w, blk, sizeof(blk), 8192) == NGX_OK);
    errno = 0;
    assert(brix_vfs_writer_spill_put(&w, blk, sizeof(blk), 8192 + 100)
           == NGX_ERROR);
    assert(errno == EINVAL);
    assert(w.mode == BRIX_VFS_WRITER_SPILL && "client error, not capacity");
    assert(g_staged_aborts == 0);
    brix_vfs_writer_spill_discard(&w);
}

static void
test_error_drain_refuses_holes_untouched_staged(void)
{
    static const u_char blk[4096];
    brix_vfs_writer_t   w;

    reset_spies();
    g_spill_root = g_root_dir;
    writer_init(&w, 0);
    assert(brix_vfs_writer_spill_enter(&w, 8192, sizeof(blk)) == NGX_OK);
    assert(brix_vfs_writer_spill_put(&w, blk, sizeof(blk), 0) == NGX_OK);
    assert(brix_vfs_writer_spill_put(&w, blk, sizeof(blk), 8192) == NGX_OK);
    errno = 0;                               /* [4096,8192) never sent */
    assert(brix_vfs_writer_spill_drain(&w) == NGX_ERROR);
    assert(errno == EINVAL);
    assert(g_staged_writes == 0 && "a hole must never be zero-filled through");
    assert(g_metric_refused == 1);
    brix_vfs_writer_spill_discard(&w);
}

/* ---- security-negative ---------------------------------------------------- */

static void
test_secneg_spill_max_at_entry_aborts_staged(void)
{
    brix_vfs_writer_t w;

    reset_spies();
    g_spill_root = g_root_dir;
    g_spill_max  = 1 << 20;                  /* brix_vfs_spill_max 1m */
    writer_init(&w, 0);
    errno = 0;                               /* triggering extent spans 4 MiB */
    assert(brix_vfs_writer_spill_enter(&w, 4 << 20, 4096) == NGX_ERROR);
    assert(errno == ENOSPC);
    assert(w.mode == BRIX_VFS_WRITER_FAILED);
    assert(g_staged_aborts == 1 && g_abort_remove_tmp == 1);
    assert(w.spill.fd == NGX_INVALID_FILE && "refused before creating any scratch");
    assert(g_metric_refused == 1);
}

static void
test_secneg_spill_max_mid_put_leaves_no_partial_object(void)
{
    static const u_char blk[4096];
    brix_vfs_writer_t   w;
    struct stat         st;
    char                kept[PATH_MAX];

    reset_spies();
    g_spill_root = g_root_dir;
    g_spill_max  = 1 << 20;
    writer_init(&w, 0);
    assert(brix_vfs_writer_spill_enter(&w, 8192, sizeof(blk)) == NGX_OK);
    assert(brix_vfs_writer_spill_put(&w, blk, sizeof(blk), 8192) == NGX_OK);
    snprintf(kept, sizeof(kept), "%s", (const char *) w.spill.path);
    errno = 0;                               /* pushes the span past the cap */
    assert(brix_vfs_writer_spill_put(&w, blk, sizeof(blk), 2 << 20)
           == NGX_ERROR);
    assert(errno == ENOSPC);
    assert(w.mode == BRIX_VFS_WRITER_FAILED);
    assert(g_staged_aborts == 1 && g_abort_remove_tmp == 1);
    assert(stat(kept, &st) != 0 && "T4 must unlink the scratch too");
    assert(g_metric_active == 0);
}

static void
test_secneg_scratch_is_private_under_spill_root_never_export(void)
{
    brix_vfs_writer_t w;
    struct stat       st;
    size_t            rlen = strlen(g_root_dir);

    reset_spies();
    g_spill_root = g_root_dir;
    writer_init(&w, 0);
    assert(brix_vfs_writer_spill_enter(&w, 4096, 1) == NGX_OK);
    assert(scratch_on_disk(&w, &st));
    assert((st.st_mode & 07777) == 0600 && "scratch must be private");
    assert(strncmp((const char *) w.spill.path, g_root_dir, rlen) == 0
           && ((const char *) w.spill.path)[rlen] == '/'
           && "scratch must live under the SERVICE spill root");
    assert(strncmp((const char *) w.spill.path, g_export_dir,
                   strlen(g_export_dir)) != 0
           && "scratch must never land under the export");
    assert(strstr((const char *) w.spill.path, ".xrd-tmp.") != NULL
           && "an orphan must be reclaimable by the owned-temp reaper");
    brix_vfs_writer_spill_discard(&w);
}

int
main(void)
{
    const char *tmp = getenv("TMPDIR");

    snprintf(g_root_dir, sizeof(g_root_dir), "%s/spillunit.XXXXXX",
             tmp != NULL && tmp[0] != '\0' ? tmp : "/tmp");
    assert(mkdtemp(g_root_dir) != NULL);
    snprintf(g_export_dir, sizeof(g_export_dir), "%s/export", g_root_dir);
    g_ctx.root_canon    = g_export_dir;      /* the ctx's EXPORT, distinct */
    g_ctx.metrics_proto = BRIX_PROTO_ROOT;
    g_pool = ngx_create_pool(4096, &g_log);
    assert(g_pool != NULL);

    test_success_enter_creates_scratch_and_promotes();
    test_success_reverse_order_drain_is_byte_exact();
    test_success_discard_unlinks_and_is_idempotent();
    test_error_no_spill_root_is_enospc_stays_sequential();
    test_error_rewind_below_staged_prefix_is_einval();
    test_error_overlap_is_einval_stays_spill();
    test_error_drain_refuses_holes_untouched_staged();
    test_secneg_spill_max_at_entry_aborts_staged();
    test_secneg_spill_max_mid_put_leaves_no_partial_object();
    test_secneg_scratch_is_private_under_spill_root_never_export();

    ngx_destroy_pool(g_pool);
    (void) rmdir(g_root_dir);
    printf("vfs_writer_spill: 10 cases OK\n");
    return 0;
}

File: tests/test_vfs_writer_spill.py
"""Phase-107 C1 — the writer's reorder spill, live over both staged-only drivers.

The VFS writer used to refuse any staged write whose offset was not the staged
cursor (`EINVAL`, pre-W2 `vfs_writer.c`).  W2 replaced the refusal with a local
spill: out-of-order extents land in a private scratch file under the export's
registered spill root, and commit drains them into the staged session in
strictly sequential order.  These are the five contract rows from
docs/refactor/phase-107-vfs-mutation-surface-completion.md §4/C1:

  success   reverse-order kXR_write over `remote` and `http` fronts lands the
            object byte-exact (the scratch existed mid-flight, and was
            unlinked by the commit);
  success   in-order writes never create a spill file at all;
  error     `brix_vfs_spill_max 1m` + a 4 MiB reordered upload refuses
            `ENOSPC` -> kXR_NoSpace, and publishes NOTHING;
  sec-neg   a read-only front refuses the write open BEFORE any spill file is
            created — asserted by listing the spill root, not only the errno;
  sec-neg   `brix_vfs_spill_path` inside an export root (plus the relative
            path and the sub-1m cap) is refused at `nginx -t`.

The C object unit (tests/c/test_vfs_writer_spill.c, `vfs_writer_spill` in
cmdscripts/c_object_units.py) proves the state machine's edges hermetically;
this file proves the wire-to-origin composition.  Instance topology in
tests/configs/nginx_p107_spill.conf.
"""
import hashlib
import os
import pathlib
import subprocess
import tempfile

import pytest

from XRootD import client
from XRootD.client.flags import OpenFlags

from _xrdcl_proxy import real_bindings_available
from fleet_lifecycle_ports import SHARED_PARSE_PLACEHOLDER_PORT
from server_launcher import LifecycleHarness, NginxInstanceSpec
from settings import BIND_HOST, HOST, NGINX_BIN

pytestmark = [pytest.mark.timeout(180),
              pytest.mark.uses_lifecycle_harness,
              pytest.mark.xdist_group("lc-p107-spill"),
              pytest.mark.skipif(
                  not real_bindings_available(),
                  reason="real libXrdCl bindings unavailable")]

SPEC = "lc-p107-spill"
MIB = 1024 * 1024
S3_AK = "AKIDP107SPILLTEST1"
S3_SK = "cDEwNy1zcGlsbC13cml0ZXItcmVvcmRlci1zZWNyZXQ"


def _bytes(n: int, seed: int) -> bytes:
    """Deterministic, non-repeating payload: a chunk-wise constant pattern
    would let a drain that permutes whole chunks pass unnoticed."""
    return bytes(((i * 37) + (i >> 12) * 11 + seed) & 0xFF for i in range(n))


def _files_under(root) -> list:
    return sorted(str(p) for p in pathlib.Path(root).rglob("*") if p.is_file())


def _need_nginx() -> None:
    if not os.access(NGINX_BIN, os.X_OK):
        pytest.skip(f"nginx binary not executable: {NGINX_BIN}")


@pytest.fixture(scope="module")
def spill_srv(tmp_path_factory):
    """One instance, four fronts + two origins (see nginx_p107_spill.conf)."""
    _need_nginx()
    base = tmp_path_factory.mktemp("p107-spill")
    dirs = {name: base / name for name in (
        "s3store", "http_origin", "exp_remote", "exp_http", "exp_capped",
        "exp_ro", "spill_remote", "spill_http", "spill_capped", "spill_ro")}
    for d in dirs.values():
        d.mkdir()
    (dirs["s3store"] / "testbucket").mkdir()
    harness = LifecycleHarness()
    try:
        ep = harness.start(NginxInstanceSpec(
            name=SPEC,
            template="nginx_p107_spill.conf",
            protocol="root",
            data_root=str(dirs["http_origin"]),
            template_values={
                "BIND_HOST": BIND_HOST,
                "S3_DIR": str(dirs["s3store"]),
                "S3_ACCESS_KEY": S3_AK,
                "S3_SECRET_KEY": S3_SK,
                "HTTP_ORIGIN_ROOT": str(dirs["http_origin"]),
                "REMOTE_EXPORT": str(dirs["exp_remote"]),
                "HTTP_EXPORT": str(dirs["exp_http"]),
                "CAPPED_EXPORT": str(dirs["exp_capped"]),
                "RO_EXPORT": str(dirs["exp_ro"]),
                "SPILL_REMOTE": str(dirs["spill_remote"]),
                "SPILL_HTTP": str(dirs["spill_http"]),
                "SPILL_CAPPED": str(dirs["spill_capped"]),
                "SPILL_RO": str(dirs["spill_ro"]),
            },
            reason="phase-107 C1 reorder spill over http + s3 staged fronts"))
        yield {"port": ep.port, "extras": ep.extra_ports, "dirs": dirs}
    finally:
        harness.close()


def _url(port: int, path: str) -> str:
    return f"root://{HOST}:{port}/{path.lstrip('/')}"


def _reverse_upload(port: int, name: str, data: bytes, spill_dir,
                    chunk: int = MIB) -> None:
    """Write `data` in strictly reverse chunk order; assert the scratch file
    appears once out-of-order bytes exist and is gone after the commit."""
    f = client.File()
    status, _ = f.open(_url(port, name), OpenFlags.DELETE | OpenFlags.NEW)
    assert status.ok, f"open for upload failed: {status.message}"
    offsets = list(range(0, len(data), chunk))[::-1]
    for i, off in enumerate(offsets):
        status, _ = f.write(data[off:off + chunk], off)
        assert status.ok, (
            f"reordered write at {off} refused: {status.message}")
        if i == 1:
            # Two out-of-order extents are in flight: the spill scratch must
            # exist NOW (owned-temp named), or the bytes went somewhere else.
            names = _files_under(spill_dir)
            assert names and all(".xrd-tmp." in n for n in names), (
                f"no owned-temp spill scratch under {spill_dir}: {names}")
    status, _ = f.close()
    assert status.ok, f"commit close failed: {status.message}"
    assert _files_under(spill_dir) == [], (
        "spill scratch survived the commit — the drain must unlink it")


def _download(port: int, name: str) -> bytes:
    f = client.File()
    status, _ = f.open(_url(port, name), OpenFlags.READ)
    assert status.ok, f"open for read failed: {status.message}"
    try:
        chunks, off = [], 0
        while True:
            status, data = f.read(offset=off, size=8 * MIB)
            assert status.ok, f"read failed: {status.message}"
            if not data:
                break
            chunks.append(bytes(data))
            off += len(data)
        return b"".join(chunks)
    finally:
        f.close()


# --------------------------------------------------------------------------- #
# success                                                                      #
# --------------------------------------------------------------------------- #
def test_reverse_order_http_is_byte_exact(spill_srv):
    """8 MiB in reverse 1 MiB chunks over the http (sd_http) front."""
    data = _bytes(8 * MIB, 3)
    _reverse_upload(spill_srv["port"], "/rev8.bin", data,
                    spill_srv["dirs"]["spill_http"])
    got = _download(spill_srv["port"], "/rev8.bin")
    assert hashlib.md5(got).hexdigest() == hashlib.md5(data).hexdigest(), (
        "reverse-order upload over http is not byte-identical")
    # The staged commit PUT the whole object to the WebDAV posix origin.
    landed = spill_srv["dirs"]["http_origin"] / "rev8.bin"
    assert landed.read_bytes() == data


def test_reverse_order_remote_is_byte_exact(spill_srv):
    """The same contract over the remote (s3://, sd_remote) driver.  Byte-
    exactness is asserted at the S3 store itself: the brix_s3-over-posix
    origin stores objects FLAT under its root (nginx_root_s3_staged.conf
    precedent), and the commit is a single signed PUT of the drained bytes."""
    port = spill_srv["extras"]["REMOTE_PORT"]
    data = _bytes(4 * MIB, 11)
    _reverse_upload(port, "/rev4.bin", data,
                    spill_srv["dirs"]["spill_remote"])
    landed = spill_srv["dirs"]["s3store"] / "rev4.bin"
    assert landed.exists(), "the staged commit published nothing to the store"
    assert hashlib.md5(landed.read_bytes()).hexdigest() \
        == hashlib.md5(data).hexdigest(), (
        "reverse-order upload over remote/s3 is not byte-identical")


def test_in_order_writes_never_create_a_spill(spill_srv):
    """Sequential writes must stay in SEQUENTIAL mode: no scratch, ever."""
    data = _bytes(4 * MIB, 5)
    spill_dir = spill_srv["dirs"]["spill_http"]
    f = client.File()
    status, _ = f.open(_url(spill_srv["port"], "/seq4.bin"),
                       OpenFlags.DELETE | OpenFlags.NEW)
    assert status.ok, f"open for upload failed: {status.message}"
    for off in range(0, len(data), MIB):
        status, _ = f.write(data[off:off + MIB], off)
        assert status.ok, f"in-order write at {off} failed: {status.message}"
        assert _files_under(spill_dir) == [], (
            "an in-order write created a spill scratch")
    status, _ = f.close()
    assert status.ok, f"close failed: {status.message}"
    landed = spill_srv["dirs"]["http_origin"] / "seq4.bin"
    assert landed.read_bytes() == data


# --------------------------------------------------------------------------- #
# error                                                                        #
# --------------------------------------------------------------------------- #
def test_spill_max_refuses_enospc_and_publishes_nothing(spill_srv):
    """brix_vfs_spill_max 1m + a 4 MiB reorder: ENOSPC -> kXR_NoSpace, the
    scratch is unlinked (T4) and NO object reaches the origin."""
    port = spill_srv["extras"]["CAPPED_PORT"]
    spill_dir = spill_srv["dirs"]["spill_capped"]
    f = client.File()
    status, _ = f.open(_url(port, "/cap4.bin"),
                       OpenFlags.DELETE | OpenFlags.NEW)
    assert status.ok, f"open for upload failed: {status.message}"
    # First extent at 3 MiB: the spill would have to reserve a 4 MiB span,
    # four times the cap — refused at entry, before any staged byte exists.
    status, _ = f.write(_bytes(MIB, 9), 3 * MIB)
    assert not status.ok, "a reorder past brix_vfs_spill_max was accepted"
    assert "space" in status.message.lower(), (
        f"want the kXR_NoSpace/ENOSPC text, got: {status.message}")
    f.close()
    assert _files_under(spill_dir) == [], (
        "the refused spill left a scratch file behind")
    assert not (spill_srv["dirs"]["http_origin"] / "cap4.bin").exists(), (
        "a refused reordered upload still published an object")


# --------------------------------------------------------------------------- #
# security-negative                                                            #
# --------------------------------------------------------------------------- #
def test_read_only_front_reaches_no_spill(spill_srv):
    """The phase-105 gate refuses BEFORE the writer exists: the spill root
    must stay empty no matter what a client attempts."""
    port = spill_srv["extras"]["RO_PORT"]
    f = client.File()
    status, _ = f.open(_url(port, "/ro.bin"), OpenFlags.DELETE | OpenFlags.NEW)
    if status.ok:  # an open that slipped through must still refuse the write
        status, _ = f.write(_bytes(MIB, 13), MIB)
        f.close()
    assert not status.ok, "a write path opened on the read-only front"
    assert _files_under(spill_srv["dirs"]["spill_ro"]) == [], (
        "a read-only front created a spill scratch — the gate ran too late")


def _nginx_t(extra_stream: str) -> tuple:
    """Parse-only stream config: export {d}/exp, spill candidates beside and
    beneath it.  Placeholders: {EXP} {UNDER} {REL} {OUT}."""
    with tempfile.TemporaryDirectory() as d:
        for sub in ("logs", "exp", "exp/spill", "out"):
            os.makedirs(os.path.join(d, sub), exist_ok=True)
        body = (extra_stream
                .replace("{EXP}", d + "/exp")
                .replace("{UNDER}", d + "/exp/spill")
                .replace("{OUT}", d + "/out")
                .replace("{REL}", "rel/spill"))
        conf = os.path.join(d, "nginx.conf")
        with open(conf, "w") as fh:
            modules = [m for m in os.environ.get(
                "TEST_NGINX_LOAD_MODULES", "").split(os.pathsep) if m]
            fh.write("".join(f"load_module {m};\n" for m in modules)
                     + f"error_log {d}/logs/e.log info;\npid {d}/logs/n.pid;\n"
                     + "events {}\nstream {\n  server {\n"
                     + f"    listen {BIND_HOST}:{SHARED_PARSE_PLACEHOLDER_PORT};\n"
                     + "    brix_root on;\n    brix_auth none;\n"
                     + f"    brix_export {d}/exp;\n"
                     + f"    brix_storage_backend posix:{d}/exp;\n"
                     + "    brix_allow_write on;\n"
                     + f"    {body}\n"
                     + "  }\n}\n")
        env = dict(os.environ, ASAN_OPTIONS="detect_leaks=0")
        r = subprocess.run([NGINX_BIN, "-t", "-c", conf, "-p", d],
                           capture_output=True, text=True, timeout=30, env=env)
    return r.returncode, r.stdout + r.stderr


@pytest.mark.parametrize("directive,needle", [
    # Inside the export: service scratch reachable as export storage.
    ("brix_vfs_spill_path {UNDER};", "at or beneath export root"),
    # Relative: realpath() would resolve it against the master's cwd.
    ("brix_vfs_spill_path {REL};", "must be absolute"),
    # A cap below the directive minimum (0 stays legal = uncapped).
    ("brix_vfs_spill_path {OUT}; brix_vfs_spill_max 4096;",
     "must be 0 or at least 1m"),
])
def test_spill_config_negatives_refused_at_nginx_t(directive, needle):
    _need_nginx()
    rc, out = _nginx_t(directive)
    assert rc != 0, f"nginx -t accepted: {directive}"
    assert needle in out, f"want {needle!r} in:\n{out[-1500:]}"
